translators/src/Print/Idris.hs

```diff
@@ -47,96 +42,109 @@ printWithImport (ImportLib NatMod) m = m
+printTm (DCon t []) = pretty t
 printTm (DCon name types) = pretty name <+> hsep (map printTm types)
-printTm (Index names ty) = braces $ typeAnn (hsep $ punctuate comma $ map pretty names) (printTm ty)
 printTm (Var var) = pretty var
 printTm (Paren e) = parens $ printTm e
 printTm (Binary op e1 e2) = printTm e1 <+> printOp2 op <+> printTm e2
-printTm (Let ds expr) = 
+printTm (Let ds expr) =
   "let" <+> align (vcat (map printLocalDefn ds) <+> "in") <> line <>
   printTm expr
-printTm (If cond thn els) =
-  "if" <+> printTm cond <+> "then" <+> printTm thn <+> "else" <+> printTm els
-printTm (Where expr ds) =
-  printTm expr <> hardline <>
-  indent 4 ("where" <> vcat (map printLocalDefn ds))
 printTm (App fun args) = printTm fun <+> (fillSep (map (group . printTm) args))
 printTm (Unary o t) = parens $ printOp1 o <+> printTm t
 printTm (Lit l) = printLit l
+printTm (KCon NatT _) = natT
+printTm (KCon StringT _) = strT
+printTm (KCon VecT [s, t]) = vectT <+> printTm t <+> printTm s
+printTm (KCon VecT _) = error "Vect takes exactly two arguments"
 
 printLit :: Literal -> Doc ann
 printLit (Nat n) = pretty n
 printLit (Bool b) = pretty b
 printLit (String str) = dquotes $ pretty str
-printLit (Vec l) =  brackets $ fillSep $ punctuate comma $ map printTm l
-printLit (List l) = brackets $ fillSep $ punctuate comma $ map printTm l
+printLit (Vec l) =  brackets $ fillSep $ punctuate vcons $ map printTm l
+printLit (List l) = brackets $ fillSep $ punctuate lcons $ map printTm l
 
 printOp1 :: Op1 -> Doc ann
 printOp1 Suc = "S"
 
 printOp2 :: Op2 -> Doc ann
 printOp2 Plus = "+"
 
+printFieldT :: FieldT -> Doc ann
+printFieldT (FieldT fname ftype) = typeAnn (pretty fname) (printTm ftype)
+
+printFieldDecl :: FieldDecl -> Doc ann
+printFieldDecl (FieldDecl fields) = vsep $ map printFieldT fields
+
+printDataConst :: Parameters -> DataCons -> Doc ann
+printDataConst p (DataCons l) = vsep $ map (prettyCon p) l
+
+prettyCon :: Parameters -> Constr -> Doc ann
+prettyCon [] (Constr n t) = typeAnn (pretty n) (printTm t)
+prettyCon p  (Constr n t) = typeAnn (pretty n)
+                                    (encloseSep emptyDoc (space <> arr) (space <> arr <> space)
+                                       (map (pretty.arg) p))
+                            <+> printTm t
+
+printCase :: Name -> Pat -> Doc ann
+printCase var (Pat a e) = pretty var <+> (hsep $ map (pretty . arg) a) <+> assign <+> printTm e
+
+printMatch :: Name -> Patterns -> Doc ann
+printMatch nm (Patterns p) = vsep (map (printCase nm) p)
+
 printLocalDefn :: LocalDefn -> Doc ann
-printLocalDefn (LocDefFun var ty args expr) = 
+printLocalDefn (LocDefFun var ty args expr) =
   typeSig <> tvar <+> assign <+> align (printTm expr)
     where
         typeSig = case ty of
-            Just t -> typeAnn (pretty var) (printTm t) <> line
+            Just t -> typeAnn (pretty var) ((hsep $ punctuate (space <> arr) $ map (printTm . argty) args) <+> arr <+> printTm t) <> line
```

Comment:
This seems like a common pattern, maybe factor it out?

Reply:
which part seems common? `hsep $ punctutate` ?